Add analytics events to Prompts Screen

metalwihen opened the pull request:
Fixes #17125 

This PR introduces analytics event tracking to the Prompts List Screen.

**To test:**
- Follow test instructions as explained in previous PRs: #2, #5, #6
- Open the Prompts Screen to see if event `blogging_prompts_list_accessed` is tracked
- Click on the different tabs of the Prompts Screen to see if event `blogging_prompts_list_tab_changed` is tracked.

**Screenshots**
Attaching screenshots of the logs that confirm that the events are getting logged:

![image](https://user-images.githubusercontent.com/2140539/195433186-35023725-c8a0-4383-aa4e-73e80f3f80c5.png)
![image](https://user-images.githubusercontent.com/2140539/195433212-562c55ca-74fd-4291-ad16-7e52960ee9a2.png)


## Regression Notes
1. Potential unintended areas of impact
- Changes were made only to the Prompts Screen 

2. What I did to test those areas of impact (or what existing automated tests I relied on)
- Run existing unit tests
- Run manual tests around the Prompts screen to see if flow is working as expected

3. What automated tests I added (or what prevented me from doing so)
- Add unit tests to test AnalyticsTracker and ViewModel

PR submission checklist:

- [x] I have completed the Regression Notes.
- [x] I have considered adding accessibility improvements for my changes.
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

metalwihen committed:
Add analytics events to Prompts Screen

RenanLukas commented on WordPress/src/main/java/org/wordpress/android/ui/bloggingprompts/BloggingPromptsAnalyticsTracker.kt:
```diff
@@ -0,0 +1,28 @@
+class BloggingPromptsAnalyticsTracker @Inject constructor(
+    private val analyticsTracker: AnalyticsTrackerWrapper
+) {
+    fun trackScreenAccessed(site: SiteModel, currentTab: PromptSection) =
```
We have examples of `trackScreenViewed` and `trackScreenShown` method names in other flows. Considering this, would it make sense to use one of the existing examples of names? 

e.g. `trackScreenViewed` or `trackScreenShown`

metalwihen replied:
My reasoning was to match the method name with the analytics event name. 

That said, I'll keep it consistent and make the changes.

metalwihen replied:
Done.

RenanLukas commented on WordPress/src/main/java/org/wordpress/android/ui/bloggingprompts/BloggingPromptsAnalyticsTracker.kt:
```diff
@@ -0,0 +1,28 @@
+    fun trackScreenAccessed(site: SiteModel, currentTab: PromptSection) =
+            analyticsTracker.track(
+                    BLOGGING_PROMPTS_LIST_ACCESSED,
+                    site,
```
I believe we don't need to add the site details for this event. What do you think?

metalwihen replied:
I believe prompts are fetched for a _site_ (and there are prompts only if a _site_ has a min number of posts). If the successful loading of the prompts page depends on the _site_, then it made sense to me to track it as a relevant property. This could be useful during debugging purposes too. 

If you feel this property is still unnecessary, I'll remove it. Could you confirm?

metalwihen replied:
I believe prompts are fetched for a _site_ (and there are prompts only if a _site_ has a min number of posts). If the successful loading of the prompts page depends on the _site_, then it made sense to me to track it as a relevant property. This could be useful during debugging purposes too. 

If you feel this property is still unnecessary, I'll remove it. Could you confirm?

metalwihen replied:
I believe prompts are fetched for a _site_ (and there are prompts only if a _site_ has a min number of posts). If the successful loading of the prompts page depends on the _site_, then it made sense to me to track it as a relevant property. This could be useful during debugging purposes too. 

If you feel this property is still unnecessary, I'll remove it. Could you confirm?

RenanLukas replied:
> This could be useful during debugging purposes too.

I agree, good idea. Let's leave it, then 👍

RenanLukas commented on WordPress/src/main/java/org/wordpress/android/ui/bloggingprompts/BloggingPromptsAnalyticsTracker.kt:
```diff
@@ -0,0 +1,28 @@
+    fun trackTabSelected(site: SiteModel, currentTab: PromptSection) =
+            analyticsTracker.track(
+                    BLOGGING_PROMPTS_LIST_TAB_CHANGED,
+                    site,
```
I believe we don't need to add the site details for this event. What do you think?

metalwihen replied:
Same response as https://github.com/metalwihen/WordPress-Android/pull/7#discussion_r994077937

metalwihen replied:
Same response as https://github.com/metalwihen/WordPress-Android/pull/7#discussion_r994077937

RenanLukas replied:
Same as https://github.com/metalwihen/WordPress-Android/pull/7#discussion_r995080003

RenanLukas commented on WordPress/src/main/java/org/wordpress/android/ui/bloggingprompts/BloggingPromptsParentFragment.kt:
```diff
@@ -76,7 +80,7 @@ class BloggingPromptsParentFragment : Fragment() {
 }
 
-private val promptsSections = listOf(ALL, ANSWERED, NOT_ANSWERED)
+internal val promptsSections = listOf(ALL, ANSWERED, NOT_ANSWERED)
```
Would it make sense to have this logic in the `ViewModel` instead? If we do that we could also unit test the order of the sections, so if someone change it by accident CI would fail.

metalwihen replied:
I missed testing the order of the tabs. That's a nice idea. Will do.

Also, yes, will move to View Model.

metalwihen replied:
Done.

RenanLukas commented on WordPress/src/main/java/org/wordpress/android/ui/bloggingprompts/BloggingPromptsParentViewModel.kt:
```diff
@@ -1,10 +1,31 @@
-    fun onOpen() = Unit
+@HiltViewModel
+class BloggingPromptsParentViewModel @Inject constructor(
+    private val handle: BloggingPromptsSiteProvider,
```
Would it make sense to call this variable `provider`, `bloggingPromptsSiteProvider` or something similar? Or to rename `BloggingPromptsSiteProvider` to something like `BloggingPromptsSiteStateHandle` or similar?

metalwihen replied:
Ah. Yes. Will do. Missed this during some refactoring.

metalwihen replied:
Ah. Yes. Will do. Missed this during some refactoring.

metalwihen replied:
Done.

RenanLukas commented on WordPress/src/main/java/org/wordpress/android/ui/bloggingprompts/BloggingPromptsParentViewModel.kt:
```diff
@@ -1,10 +1,31 @@
+    }
+
+    @VisibleForTesting
+    internal fun getSite(): SiteModel = checkNotNull(handle.getSite()) {
```
I couldn't find this method being called directly in the unit tests. If so, would it make sense to make it private and remove the `@VisibleForTesting` annotation? 🤔 

I'm thinking that maybe mocking the return of `BloggingPromptsSiteProvider` would be enough for unit tests, but I might be wrong here.

RenanLukas replied:
Besides that, as far as I understand `checkNotNull` will throw a `RuntimeException` if the value is `null`. If that's correct, I suppose is going to crash the app, right? Would it make sense to handle it differently, like showing the error state for example?

metalwihen replied:
> I couldn't find this method being called directly in the unit tests. If so, would it make sense to make it private and remove the @VisibleForTesting annotation? 🤔

Another miss when I refactored the tests later. Will make the changes. 

> handle it differently, like showing the error state for example?

Agreed. I'm a little embarrassed I let the crash happen.

That makes me wonder though. **Should we show an error screen (blocking the user if there is analytics related issues) or let it silently fail with logging (ensuring the feature can still be accessed)?** We can choose to either compromise on the accuracy of analytics events or the user experience. This is of course assuming the rest of the pages do not require the site argument which it might (leading to the same error screen). Will keep this in mind for the next PR around networking layer.

metalwihen replied:
- Removed the @VisibleForTesting  
- Removed the runtime exception. If siteModel is missing, tracking will silently fail. 

Note: I couldn't add logging because AppLog needs an enum as a TAG and I couldn't add BloggingPrompts since it's in a different library. Any suggestions?

RenanLukas replied:
> Should we show an error screen (blocking the user if there is analytics related issues) or let it silently fail with logging (ensuring the feature can still be accessed)?

> This is of course assuming the rest of the pages do not require the site argument which it might (leading to the same error screen)

Good point. I believe this will become clearer when the network layer is implemented, but IMHO we should not let the screen show an error just because of the analytics tracking. I was assuming that this `SiteModel` would be used by the main logic too, but I might be wrong 🙂 

> I couldn't add logging because AppLog needs an enum as a TAG and I couldn't add BloggingPrompts since it's in a different library. Any suggestions?

Hmm did you try something like [this](https://github.com/wordpress-mobile/WordPress-Android/blob/trunk/WordPress/src/main/java/org/wordpress/android/util/WPActivityUtils.java#L51)?

metalwihen replied:
>I was assuming that this SiteModel would be used by the main logic too

Yes, I'm assuming the same. 

>  Hmm did you try something like this

My mistake. Realized I misinterpreted an error. I can edit T enum in `org.wordpress.android.util`. On searching, I must have opened the AppLog of another module.

RenanLukas commented on WordPress/src/main/java/org/wordpress/android/ui/bloggingprompts/BloggingPromptsParentViewModel.kt:
```diff
@@ -1,10 +1,31 @@
-class BloggingPromptsParentViewModel : ViewModel() {
-    fun onOpen() = Unit
+@HiltViewModel
+class BloggingPromptsParentViewModel @Inject constructor(
```
I was wondering if it makes sense to rename this `ViewModel` and `Activity` so their names match. I was thinking about something related to the prompts list (e.g. `BloggingPromptsListViewModel` and `BloggingPromptsListActivity` or something like that 😁). What do you think? Does it make sense?

metalwihen replied:
**Current Reasoning.** 
The `BloggingPromptsParentViewModel` is used in the both `BloggingPromptsParentFragment` and `BloggingPromptsActivity`. I matched the name of the viewmodel with the name of the fragment instead of the activity. I'm also planning to introduce `BloggingPromptsListViewModel` later to handle the fetching and rendering of prompts. 

Group of Classes 1:
- BloggingPromptsParentViewModel
- BloggingPromptsParentFragment
- BloggingPromptsActivity

Group of Classes 2:
- BloggingPromptsListViewModel (To be Created) 
- BloggingPromptsListFragment

**New Approach.** 

Another approach is to share the same viewmodel across all three classes (Activity, ParentFragment, ListFragment), at which point, the name you suggested `BloggingPromptsListViewModel` might match more. This might also make it easier to handle error states as noticed in the analytics. 

Group of Classes:
- BloggingPromptsListViewModel
- BloggingPromptsParentFragment
- BloggingPromptsActivity
- BloggingPromptsListFragment

I'll go ahead with this approach then. Let me know if you feel otherwise.  Naming is tricky!

metalwihen replied:
I've renamed the classes, all having the prefix `BloggingPromptsList`. Please take a look.

RenanLukas replied:
> Naming is tricky!

It's always the hardest part 🙂 IMHO the common prefix looks good, thanks

RenanLukas commented on WordPress/src/main/java/org/wordpress/android/ui/bloggingprompts/BloggingPromptsSiteProvider.kt:
```diff
@@ -0,0 +1,14 @@
+import javax.inject.Inject
+
+class BloggingPromptsSiteProvider @Inject constructor(
+    private val savedStateHandle: SavedStateHandle
```
I haven't tried `SavedStateHandle` yet so I might be wrong: should we be using [getLiveData](https://developer.android.com/topic/libraries/architecture/viewmodel-savedstate#savedstate-livedata) instead? 🤔

metalwihen replied:
My reasoning was that it's only being used for analytics events in the view model. So livedata seemed unnecessary (from the perspective of observer-pattern). 

If needed, according to the [documentation](https://developer.android.com/topic/libraries/architecture/viewmodel-savedstate), a livedata variable can be retrieved from the SavedStateHandle. `savedStateHandle.getLiveData<String>()`

As for why I'm using SavedStateHandle over directly using a LiveData variable. According the documentation, it's more resilient in holding the values during a system initiated death of activity or fragment. It also handles config changes well (rotating the screen). 

Does that make sense?

RenanLukas replied:
Yes, makes sense 👍  thanks for clarifying

RenanLukas replied:
Last question about this: is this part of [the docs](https://developer.android.com/topic/libraries/architecture/viewmodel-savedstate#setup) relevant to this scenario?

> When providing a custom [ViewModelProvider.Factory](https://developer.android.com/reference/androidx/lifecycle/ViewModelProvider.Factory) instance, you can enable usage of SavedStateHandle by extending [AbstractSavedStateViewModelFactory](https://developer.android.com/reference/androidx/lifecycle/AbstractSavedStateViewModelFactory).

metalwihen replied:
Yes but I believe Hilt is taking care of it internally. I ran the code to make sure it's running.

Relevant documentation: https://dagger.dev/hilt/view-model.html

RenanLukas commented on WordPress/src/test/java/org/wordpress/android/ui/bloggingprompts/BloggingPromptsParentViewModelTest.kt:
```diff
@@ -0,0 +1,46 @@
+    )
+
+    @Test
+    fun `Should save siteModel when started`() {
```
Nice 👍

RenanLukas reviewed (commented):
Hey @metalwihen , thanks for the changes.

I've left a few questions. Please let me know if they make sense 👍

metalwihen committed:
Apply PR suggestions

- Rename analytics tracker method
- Move promptSection enum to ViewModel from Fragment
- Remove @VisibleForTesting for methods that aren't tested
- Add test to ensure tabs on Prompts screen is in right order

metalwihen committed:
Apply PR Suggestion: Rename Blogging Prompt List classes

RenanLukas reviewed (approved):
Thanks for the changes @metalwihen 
LGTM :shipit:
